{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00ee2077",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-07-21T12:38:50.742104Z",
     "iopub.status.busy": "2025-07-21T12:38:50.741808Z",
     "iopub.status.idle": "2025-07-21T12:38:55.069367Z",
     "shell.execute_reply": "2025-07-21T12:38:55.068956Z"
    },
    "papermill": {
     "duration": 4.331763,
     "end_time": "2025-07-21T12:38:55.070205",
     "exception": false,
     "start_time": "2025-07-21T12:38:50.738442",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61e904c4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-07-21T12:38:55.075643Z",
     "iopub.status.busy": "2025-07-21T12:38:55.075350Z",
     "iopub.status.idle": "2025-07-21T12:38:55.081841Z",
     "shell.execute_reply": "2025-07-21T12:38:55.081513Z"
    },
    "papermill": {
     "duration": 0.010126,
     "end_time": "2025-07-21T12:38:55.082384",
     "exception": false,
     "start_time": "2025-07-21T12:38:55.072258",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_gapfill_likelihoods(gapfill,error_rate_dicts,alleles):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    likelihood_list = {}\n",
    "    for allele in alleles:\n",
    "        if allele in likelihoods:\n",
    "            likelihood_list[allele] = likelihoods[allele]\n",
    "        else:\n",
    "            likelihood_list[allele] = 10e-10\n",
    "    return likelihood_list\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1cb36f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-07-21T12:38:55.086346Z",
     "iopub.status.busy": "2025-07-21T12:38:55.086101Z",
     "iopub.status.idle": "2025-07-21T12:38:55.088004Z",
     "shell.execute_reply": "2025-07-21T12:38:55.087705Z"
    },
    "papermill": {
     "duration": 0.004725,
     "end_time": "2025-07-21T12:38:55.088499",
     "exception": false,
     "start_time": "2025-07-21T12:38:55.083774",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "feature_set = 'feature_set_all.csv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5302573",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-07-21T12:38:55.091701Z",
     "iopub.status.busy": "2025-07-21T12:38:55.091578Z",
     "iopub.status.idle": "2025-07-21T13:00:32.128511Z",
     "shell.execute_reply": "2025-07-21T13:00:32.125709Z"
    },
    "papermill": {
     "duration": 1297.040732,
     "end_time": "2025-07-21T13:00:32.130632",
     "exception": false,
     "start_time": "2025-07-21T12:38:55.089900",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1756198 UMIs found\n",
      "Collapsing UMIs across probes, 1756198 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1749625 UMIs remaining (99.63%)\n",
      "Filtering cells based on min counts (0) and genes (0) in WTA\n",
      "Filtering probe reads based on cell barcodes in adata, 1642071 UMIs remaining (93.50%)\n",
      "2010873 UMIs found\n",
      "Collapsing UMIs across probes, 2010873 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2003477 UMIs remaining (99.63%)\n",
      "Filtering cells based on min counts (0) and genes (0) in WTA\n",
      "Filtering probe reads based on cell barcodes in adata, 1875420 UMIs remaining (93.26%)\n"
     ]
    }
   ],
   "source": [
    "for lib in ['1plex','1','2','4plex']:\n",
    "    if lib == '4plex':\n",
    "        BCs = ['BC001', 'BC002', 'BC003', 'BC004']\n",
    "    elif lib == '1plex':\n",
    "        BCs = ['BC001']\n",
    "    else:\n",
    "        BCs = ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']\n",
    "\n",
    "    for BC in BCs:\n",
    "        gf_dirs = {}\n",
    "\n",
    "        ### first get probe_reads to use for the patient\n",
    "        if lib == '4plex':\n",
    "            adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250318_rerun_JAK2_patients_' + BC + '.h5ad'\n",
    "            gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_4plex_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "            gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_4plex_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "        elif lib == '1plex':\n",
    "            adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20241114_rerun_JAK2_patients_' + BC + '.h5ad'\n",
    "            gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_20241114/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "        else:\n",
    "            adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "            gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "            gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "\n",
    "        min_percent_supporting = 0.9\n",
    "        collapse_across_probes = True\n",
    "        for key,gf_dir in gf_dirs.items():\n",
    "            if key == 0:\n",
    "                probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "            else:\n",
    "                probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "        sample = BC + '_' + lib\n",
    "\n",
    "        alt_gapfill_set = pd.read_csv(feature_set)\n",
    "\n",
    "        sample = BC + '_' + lib\n",
    "        alt_gapfill_set = alt_gapfill_set.loc[alt_gapfill_set['sample'] == sample]\n",
    "\n",
    "        if len(alt_gapfill_set) > 0:\n",
    "            ## reformat alt gapfill set\n",
    "            manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "            probe_name_to_probe_idx = dict(zip(manifest['name'], manifest['index']))\n",
    "            alt_gapfill_set['probe_idx'] = alt_gapfill_set['original_name'].map(probe_name_to_probe_idx)\n",
    "            \n",
    "            alt_gapfill_set = pd.concat([\n",
    "                alt_gapfill_set[['probe_idx', 'name', 'gapfill']].rename(columns={'gapfill': 'gapfill_value'}).assign(variant=True),\n",
    "                alt_gapfill_set[['probe_idx', 'name', 'gapfill_from_transcriptome']].rename(columns={'gapfill_from_transcriptome': 'gapfill_value'}).assign(variant=False)\n",
    "            ], ignore_index=True)\n",
    "\n",
    "            alt_gapfill_set = alt_gapfill_set.dropna(subset='probe_idx')\n",
    "            # print(alt_gapfill_set)\n",
    "            \n",
    "            ### from now on, we only need to consider probes that have an alt gapfill\n",
    "            probe_reads = probe_reads.loc[probe_reads['probe_idx'].isin(alt_gapfill_set['probe_idx'])]\n",
    "\n",
    "            unique_gapfill_list = probe_reads[['probe_idx','gapfill']].drop_duplicates()\n",
    "\n",
    "            possible_alleles = {}\n",
    "            for probe_idx in probe_reads['probe_idx'].unique():\n",
    "                possible_alleles[probe_idx] = alt_gapfill_set.loc[alt_gapfill_set['probe_idx'] == probe_idx, 'gapfill_value'].unique()\n",
    "\n",
    "            unique_gapfill_list['likelihood'] = unique_gapfill_list.apply(\n",
    "                lambda row: get_gapfill_likelihoods(row['gapfill'], error_rate_dicts, possible_alleles[row['probe_idx']]), axis=1\n",
    "            )\n",
    "\n",
    "            for probe_idx in possible_alleles.keys():\n",
    "                likelihood_keys = possible_alleles[probe_idx]\n",
    "                for key in likelihood_keys:\n",
    "                    unique_gapfill_list.loc[unique_gapfill_list['probe_idx'] == probe_idx,key] = unique_gapfill_list.loc[unique_gapfill_list['probe_idx'] == probe_idx, 'likelihood'].apply(lambda x: x.get(key, None))\n",
    "                \n",
    "            unique_gapfill_list = unique_gapfill_list.drop(columns=['likelihood'])\n",
    "\n",
    "            unique_gapfill_list.to_csv('likelihood_tables_' + lib + '/' + BC + '_gapfill_likelihoods.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2f7aef3",
   "metadata": {
    "papermill": {
     "duration": 0.009647,
     "end_time": "2025-07-21T13:00:32.153932",
     "exception": false,
     "start_time": "2025-07-21T13:00:32.144285",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1304.091679,
   "end_time": "2025-07-21T13:00:33.684308",
   "environment_variables": {},
   "exception": null,
   "input_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "output_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-07-21T12:38:49.592629",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
