{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c27a3cfc",
   "metadata": {
    "papermill": {
     "duration": 0.001947,
     "end_time": "2025-06-30T01:07:48.910192",
     "exception": false,
     "start_time": "2025-06-30T01:07:48.908245",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Below saves the frequencies and likelihoods of all gapfills for all patients. Originally this only saved non-wildtype but it is informative to keep numbers of all.\n",
    "This strategy could be generic, although likelihoods should be updated with empirical frequencies when available (i.e. if those probes have been used on normal tissue or cell lines; in this case they have not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "54b95d4a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:48.914233Z",
     "iopub.status.busy": "2025-06-30T01:07:48.914057Z",
     "iopub.status.idle": "2025-06-30T01:07:51.942164Z",
     "shell.execute_reply": "2025-06-30T01:07:51.941758Z"
    },
    "papermill": {
     "duration": 3.03121,
     "end_time": "2025-06-30T01:07:51.943193",
     "exception": false,
     "start_time": "2025-06-30T01:07:48.911983",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c693951b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.947410Z",
     "iopub.status.busy": "2025-06-30T01:07:51.947129Z",
     "iopub.status.idle": "2025-06-30T01:07:51.960090Z",
     "shell.execute_reply": "2025-06-30T01:07:51.959790Z"
    },
    "papermill": {
     "duration": 0.015396,
     "end_time": "2025-06-30T01:07:51.960622",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.945226",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_wt_likelihood(gapfill,error_rate_dicts,wt_gapfill):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    if wt_gapfill in likelihoods:\n",
    "        return likelihoods[wt_gapfill]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)\n",
    "\n",
    "os.system('mkdir unexpected_gapfill_likelihoods')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6dfbf134",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.964307Z",
     "iopub.status.busy": "2025-06-30T01:07:51.964093Z",
     "iopub.status.idle": "2025-06-30T01:07:51.965925Z",
     "shell.execute_reply": "2025-06-30T01:07:51.965630Z"
    },
    "papermill": {
     "duration": 0.004256,
     "end_time": "2025-06-30T01:07:51.966455",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.962199",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "lib = '1'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8023af43",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.970094Z",
     "iopub.status.busy": "2025-06-30T01:07:51.969958Z",
     "iopub.status.idle": "2025-06-30T01:27:06.595126Z",
     "shell.execute_reply": "2025-06-30T01:27:06.594503Z"
    },
    "papermill": {
     "duration": 1154.628246,
     "end_time": "2025-06-30T01:27:06.596177",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.967931",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "568648 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 568648 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 566830 UMIs remaining (99.68%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 553555 UMIs remaining (97.35%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "674061 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 674061 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 671947 UMIs remaining (99.69%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 655189 UMIs remaining (97.20%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387979 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 387979 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 386424 UMIs remaining (99.60%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 366951 UMIs remaining (94.58%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "447368 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 447368 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 445728 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 417432 UMIs remaining (93.31%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "616887 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 616887 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 614468 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 593738 UMIs remaining (96.25%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "720635 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 720635 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 717840 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 687066 UMIs remaining (95.34%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "917786 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 917786 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 914057 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 852264 UMIs remaining (92.86%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1111593 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1111593 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1107432 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1020593 UMIs remaining (91.81%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1419549 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1419549 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1414497 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1364100 UMIs remaining (96.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1645255 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1645255 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1639748 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1580943 UMIs remaining (96.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1994670 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1994670 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1987444 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1861145 UMIs remaining (93.31%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2261733 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2261733 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2254124 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2089567 UMIs remaining (92.39%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1756198 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1756198 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1749625 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1613208 UMIs remaining (91.86%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2010873 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2010873 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2003477 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1828215 UMIs remaining (90.92%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45584 UMIs found\n",
      "Collapsing UMIs across probes, 45584 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 45445 UMIs remaining (99.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 44990 UMIs remaining (98.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55256 UMIs found\n",
      "Collapsing UMIs across probes, 55256 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 55122 UMIs remaining (99.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 40745 UMIs remaining (73.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58081 UMIs found\n",
      "Collapsing UMIs across probes, 58081 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 57917 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 57091 UMIs remaining (98.30%)\n",
      "74056 UMIs found\n",
      "Collapsing UMIs across probes, 74056 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 73835 UMIs remaining (99.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 65548 UMIs remaining (88.51%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2292046 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2292046 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2281933 UMIs remaining (99.56%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2106774 UMIs remaining (91.92%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2796720 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2796720 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2785174 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2571310 UMIs remaining (91.94%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1138142 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1138142 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1133722 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1054189 UMIs remaining (92.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1271503 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1271503 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1266866 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1082304 UMIs remaining (85.12%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1345020 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1345020 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1339466 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1288440 UMIs remaining (95.79%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1558831 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1558831 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1552968 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1480120 UMIs remaining (94.95%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1450740 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1450740 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1445663 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1401531 UMIs remaining (96.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1651562 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1651562 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1646034 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1587987 UMIs remaining (96.15%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "139217 UMIs found\n",
      "Collapsing UMIs across probes, 139217 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 138856 UMIs remaining (99.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 132953 UMIs remaining (95.50%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179556 UMIs found\n",
      "Collapsing UMIs across probes, 179556 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 179076 UMIs remaining (99.73%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 171766 UMIs remaining (95.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1729224 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1729224 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1722106 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1611729 UMIs remaining (93.21%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2009030 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2009030 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2001147 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1857143 UMIs remaining (92.44%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1368673 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1368673 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1361869 UMIs remaining (99.50%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1298549 UMIs remaining (94.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1630738 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1630738 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1623116 UMIs remaining (99.53%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1550467 UMIs remaining (95.08%)\n"
     ]
    }
   ],
   "source": [
    "### first get probe_reads to use for the patient\n",
    "\n",
    "for BC in ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']:\n",
    "    adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "\n",
    "    gf_dirs = {}\n",
    "    gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "\n",
    "    min_percent_supporting = 0.9\n",
    "    collapse_across_probes = True\n",
    "    for key,gf_dir in gf_dirs.items():\n",
    "        if key == 0:\n",
    "            probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "        else:\n",
    "            probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "    exclude_keyword = '0bp'\n",
    "    manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "    if 'gapfill_from_transcriptome' not in probe_reads.columns:\n",
    "        probe_reads = probe_reads.merge(manifest[['gapfill_from_transcriptome','name']], left_on='probe_idx',right_index=True,how='left')\n",
    "    probe_reads['gapfill'] = probe_reads['gapfill'].fillna('')\n",
    "    probe_reads['gapfill_from_transcriptome'] = probe_reads['gapfill_from_transcriptome'].fillna('')\n",
    "    if probe_reads['name'].isna().any():\n",
    "        raise ValueError('Probes with no name found in probe reads for ' + BC)\n",
    "    probe_reads = probe_reads.loc[probe_reads['name'].str.contains(exclude_keyword) == False]\n",
    "\n",
    "    counts_in_library = len(probe_reads)\n",
    "    gapfill_frequencies = probe_reads.groupby('probe_idx')['gapfill'].value_counts(normalize=True).rename('frequency').reset_index()\n",
    "    gapfill_counts = probe_reads.groupby('probe_idx')['gapfill'].value_counts().rename('count_of_this_gapfill').reset_index()\n",
    "    probe_counts = probe_reads.groupby('probe_idx').size().rename('count_of_this_probe').reset_index()\n",
    "\n",
    "    to_resolve = probe_reads[['probe_idx','name','gapfill','gapfill_from_transcriptome']].drop_duplicates()\n",
    "    to_resolve['likelihood'] = to_resolve.apply(\n",
    "        lambda row: get_wt_likelihood(row['gapfill'], error_rate_dicts, row['gapfill_from_transcriptome']), axis=1\n",
    "    )\n",
    "    to_resolve = to_resolve.merge(gapfill_frequencies, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(gapfill_counts, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(probe_counts, on = ['probe_idx'], how='left')\n",
    "\n",
    "    to_add = manifest.loc[~(manifest.index.isin(to_resolve['probe_idx'])) & (manifest['name'].str.contains(exclude_keyword) == False)].reset_index().rename(columns={'index': 'probe_idx'})\n",
    "    to_resolve = pd.concat([to_resolve, to_add[['probe_idx','name','gapfill_from_transcriptome']]]).reset_index(drop=True)\n",
    "    to_resolve[['count_of_this_gapfill','count_of_this_probe']] = to_resolve[['count_of_this_gapfill','count_of_this_probe']].fillna(0)\n",
    "    to_resolve.to_csv('unexpected_gapfill_likelihoods/unresolved_gapfills_' + BC + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6354e4e",
   "metadata": {
    "papermill": {
     "duration": 0.00586,
     "end_time": "2025-06-30T01:27:06.609289",
     "exception": false,
     "start_time": "2025-06-30T01:27:06.603429",
     "status": "completed"
    },
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4161a58",
   "metadata": {
    "papermill": {
     "duration": 0.005621,
     "end_time": "2025-06-30T01:27:06.620298",
     "exception": false,
     "start_time": "2025-06-30T01:27:06.614677",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1159.706566,
   "end_time": "2025-06-30T01:27:07.641859",
   "environment_variables": {},
   "exception": null,
   "input_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "output_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-06-30T01:07:47.935293",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}