{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "39b78ca9",
   "metadata": {
    "papermill": {
     "duration": 0.00213,
     "end_time": "2025-06-30T01:36:19.116509",
     "exception": false,
     "start_time": "2025-06-30T01:36:19.114379",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Below saves the frequencies and likelihoods of all gapfills for all patients. Originally this only saved non-wildtype but it is informative to keep numbers of all.\n",
    "This strategy could be generic, although likelihoods should be updated with empirical frequencies when available (i.e. if those probes have been used on normal tissue or cell lines; in this case they have not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ecf77254",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:36:19.120463Z",
     "iopub.status.busy": "2025-06-30T01:36:19.120278Z",
     "iopub.status.idle": "2025-06-30T01:36:22.476127Z",
     "shell.execute_reply": "2025-06-30T01:36:22.475732Z"
    },
    "papermill": {
     "duration": 3.359081,
     "end_time": "2025-06-30T01:36:22.477331",
     "exception": false,
     "start_time": "2025-06-30T01:36:19.118250",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1837651c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:36:22.482340Z",
     "iopub.status.busy": "2025-06-30T01:36:22.482084Z",
     "iopub.status.idle": "2025-06-30T01:36:22.493738Z",
     "shell.execute_reply": "2025-06-30T01:36:22.493438Z"
    },
    "papermill": {
     "duration": 0.014096,
     "end_time": "2025-06-30T01:36:22.494283",
     "exception": false,
     "start_time": "2025-06-30T01:36:22.480187",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_wt_likelihood(gapfill,error_rate_dicts,wt_gapfill):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    if wt_gapfill in likelihoods:\n",
    "        return likelihoods[wt_gapfill]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)\n",
    "\n",
    "os.system('mkdir unexpected_gapfill_likelihoods')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "86cffd6f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:36:22.497887Z",
     "iopub.status.busy": "2025-06-30T01:36:22.497621Z",
     "iopub.status.idle": "2025-06-30T01:36:22.499470Z",
     "shell.execute_reply": "2025-06-30T01:36:22.499174Z"
    },
    "papermill": {
     "duration": 0.004156,
     "end_time": "2025-06-30T01:36:22.499986",
     "exception": false,
     "start_time": "2025-06-30T01:36:22.495830",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "lib = '2'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa309c8f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:36:22.503437Z",
     "iopub.status.busy": "2025-06-30T01:36:22.503290Z",
     "iopub.status.idle": "2025-06-30T01:52:27.383748Z",
     "shell.execute_reply": "2025-06-30T01:52:27.383143Z"
    },
    "papermill": {
     "duration": 964.883332,
     "end_time": "2025-06-30T01:52:27.384839",
     "exception": false,
     "start_time": "2025-06-30T01:36:22.501507",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455940 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 455940 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 454379 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 454178 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "587745 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 587745 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 586075 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 583402 UMIs remaining (99.26%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "115394 UMIs found\n",
      "Collapsing UMIs across probes, 115394 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 115015 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 114456 UMIs remaining (99.19%)\n",
      "145240 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 145240 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 144879 UMIs remaining (99.75%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 121362 UMIs remaining (83.56%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38761 UMIs found\n",
      "Collapsing UMIs across probes, 38761 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 38617 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 38465 UMIs remaining (99.24%)\n",
      "53546 UMIs found\n",
      "Collapsing UMIs across probes, 53546 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 53391 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 53049 UMIs remaining (99.07%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "353446 UMIs found\n",
      "Collapsing UMIs across probes, 353446 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 352206 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 349922 UMIs remaining (99.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "482959 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 482959 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 481617 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 468882 UMIs remaining (97.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1170446 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1170446 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1166004 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1159525 UMIs remaining (99.07%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1166210 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1166210 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1162784 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1147898 UMIs remaining (98.43%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1154939 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1154939 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1150414 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1147145 UMIs remaining (99.33%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1131641 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1131641 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1128366 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1115101 UMIs remaining (98.54%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "104353 UMIs found\n",
      "Collapsing UMIs across probes, 104353 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 103998 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 94232 UMIs remaining (90.30%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "132658 UMIs found\n",
      "Collapsing UMIs across probes, 132658 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 132336 UMIs remaining (99.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 122885 UMIs remaining (92.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93523 UMIs found\n",
      "Collapsing UMIs across probes, 93523 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 93165 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 92716 UMIs remaining (99.14%)\n",
      "141999 UMIs found\n",
      "Collapsing UMIs across probes, 141999 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 141627 UMIs remaining (99.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 137261 UMIs remaining (96.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59872 UMIs found\n",
      "Collapsing UMIs across probes, 59872 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 59669 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 52090 UMIs remaining (87.00%)\n",
      "109965 UMIs found\n",
      "Collapsing UMIs across probes, 109965 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 109655 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 94662 UMIs remaining (86.08%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "165877 UMIs found\n",
      "Collapsing UMIs across probes, 165877 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 165261 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 127083 UMIs remaining (76.61%)\n",
      "209341 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 209341 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 208749 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 177756 UMIs remaining (84.91%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1685834 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1685834 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1677744 UMIs remaining (99.52%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1412024 UMIs remaining (83.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1603023 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1603023 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1597109 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1477917 UMIs remaining (92.20%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "402426 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 402426 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 400782 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 356506 UMIs remaining (88.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "480812 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 480812 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 479242 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 441774 UMIs remaining (91.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2108039 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2108039 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2098022 UMIs remaining (99.52%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1705153 UMIs remaining (80.89%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2082358 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2082358 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2074373 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1919410 UMIs remaining (92.17%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "156687 UMIs found\n",
      "Collapsing UMIs across probes, 156687 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 156108 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 139103 UMIs remaining (88.78%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273539 UMIs found\n",
      "Collapsing UMIs across probes, 273539 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 272799 UMIs remaining (99.73%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 254066 UMIs remaining (92.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2246870 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2246870 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2236718 UMIs remaining (99.55%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2230276 UMIs remaining (99.26%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2207024 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2207024 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2199227 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2181996 UMIs remaining (98.87%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2102281 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2102281 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2091280 UMIs remaining (99.48%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2090496 UMIs remaining (99.44%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2095903 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2095903 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2087422 UMIs remaining (99.60%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2076505 UMIs remaining (99.07%)\n"
     ]
    }
   ],
   "source": [
    "### first get probe_reads to use for the patient\n",
    "\n",
    "for BC in ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']:\n",
    "    adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "\n",
    "    gf_dirs = {}\n",
    "    gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "\n",
    "    min_percent_supporting = 0.9\n",
    "    collapse_across_probes = True\n",
    "    for key,gf_dir in gf_dirs.items():\n",
    "        if key == 0:\n",
    "            probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "        else:\n",
    "            probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "    exclude_keyword = '0bp'\n",
    "    manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "    if 'gapfill_from_transcriptome' not in probe_reads.columns:\n",
    "        probe_reads = probe_reads.merge(manifest[['gapfill_from_transcriptome','name']], left_on='probe_idx',right_index=True,how='left')\n",
    "    probe_reads['gapfill'] = probe_reads['gapfill'].fillna('')\n",
    "    probe_reads['gapfill_from_transcriptome'] = probe_reads['gapfill_from_transcriptome'].fillna('')\n",
    "    if probe_reads['name'].isna().any():\n",
    "        raise ValueError('Probes with no name found in probe reads for ' + BC)\n",
    "    probe_reads = probe_reads.loc[probe_reads['name'].str.contains(exclude_keyword) == False]\n",
    "\n",
    "    counts_in_library = len(probe_reads)\n",
    "    gapfill_frequencies = probe_reads.groupby('probe_idx')['gapfill'].value_counts(normalize=True).rename('frequency').reset_index()\n",
    "    gapfill_counts = probe_reads.groupby('probe_idx')['gapfill'].value_counts().rename('count_of_this_gapfill').reset_index()\n",
    "    probe_counts = probe_reads.groupby('probe_idx').size().rename('count_of_this_probe').reset_index()\n",
    "\n",
    "    to_resolve = probe_reads[['probe_idx','name','gapfill','gapfill_from_transcriptome']].drop_duplicates()\n",
    "    to_resolve['likelihood'] = to_resolve.apply(\n",
    "        lambda row: get_wt_likelihood(row['gapfill'], error_rate_dicts, row['gapfill_from_transcriptome']), axis=1\n",
    "    )\n",
    "    to_resolve = to_resolve.merge(gapfill_frequencies, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(gapfill_counts, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(probe_counts, on = ['probe_idx'], how='left')\n",
    "\n",
    "    to_add = manifest.loc[~(manifest.index.isin(to_resolve['probe_idx'])) & (manifest['name'].str.contains(exclude_keyword) == False)].reset_index().rename(columns={'index': 'probe_idx'})\n",
    "    to_resolve = pd.concat([to_resolve, to_add[['probe_idx','name','gapfill_from_transcriptome']]]).reset_index(drop=True)\n",
    "    to_resolve[['count_of_this_gapfill','count_of_this_probe']] = to_resolve[['count_of_this_gapfill','count_of_this_probe']].fillna(0)\n",
    "    to_resolve.to_csv('unexpected_gapfill_likelihoods/unresolved_gapfills_' + BC + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97697564",
   "metadata": {
    "papermill": {
     "duration": 0.00684,
     "end_time": "2025-06-30T01:52:27.400243",
     "exception": false,
     "start_time": "2025-06-30T01:52:27.393403",
     "status": "completed"
    },
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72f975df",
   "metadata": {
    "papermill": {
     "duration": 0.006807,
     "end_time": "2025-06-30T01:52:27.413643",
     "exception": false,
     "start_time": "2025-06-30T01:52:27.406836",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 970.648185,
   "end_time": "2025-06-30T01:52:28.536662",
   "environment_variables": {},
   "exception": null,
   "input_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "output_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-06-30T01:36:17.888477",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}