{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "68a7af20",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T02:08:40.099878Z",
     "iopub.status.busy": "2025-06-30T02:08:40.099660Z",
     "iopub.status.idle": "2025-06-30T02:08:43.186110Z",
     "shell.execute_reply": "2025-06-30T02:08:43.185702Z"
    },
    "papermill": {
     "duration": 3.090302,
     "end_time": "2025-06-30T02:08:43.187281",
     "exception": false,
     "start_time": "2025-06-30T02:08:40.096979",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e4c7a84",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T02:08:43.198519Z",
     "iopub.status.busy": "2025-06-30T02:08:43.198231Z",
     "iopub.status.idle": "2025-06-30T02:08:43.206287Z",
     "shell.execute_reply": "2025-06-30T02:08:43.205973Z"
    },
    "papermill": {
     "duration": 0.017308,
     "end_time": "2025-06-30T02:08:43.206832",
     "exception": false,
     "start_time": "2025-06-30T02:08:43.189524",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_gapfill_likelihoods(gapfill,error_rate_dicts,alleles):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    likelihood_list = {}\n",
    "    for allele in alleles:\n",
    "        if allele in likelihoods:\n",
    "            likelihood_list[allele] = likelihoods[allele]\n",
    "        else:\n",
    "            likelihood_list[allele] = 10e-10\n",
    "    return likelihood_list\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "603e50bf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T02:08:43.210438Z",
     "iopub.status.busy": "2025-06-30T02:08:43.210155Z",
     "iopub.status.idle": "2025-06-30T02:08:43.223730Z",
     "shell.execute_reply": "2025-06-30T02:08:43.223411Z"
    },
    "papermill": {
     "duration": 0.015933,
     "end_time": "2025-06-30T02:08:43.224382",
     "exception": false,
     "start_time": "2025-06-30T02:08:43.208449",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "alt_gapfill_set1 = pd.read_csv('../part1/likelihood_tables/all_gapfill_features_filtered.csv')\n",
    "alt_gapfill_set2 = pd.read_csv('../part2/likelihood_tables/all_gapfill_features_filtered.csv')\n",
    "alt_gapfill_set_combined = pd.concat([alt_gapfill_set1, alt_gapfill_set2])\n",
    "alt_gapfill_set_combined = alt_gapfill_set_combined.sort_values('likelihood_wt_given_observed').drop_duplicates(subset='name')\n",
    "\n",
    "alt_gapfill_set_combined.to_csv('likelihood_tables/all_gapfill_features_filtered_combined.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d943c2b9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T02:08:43.227752Z",
     "iopub.status.busy": "2025-06-30T02:08:43.227586Z",
     "iopub.status.idle": "2025-06-30T02:08:43.229456Z",
     "shell.execute_reply": "2025-06-30T02:08:43.229151Z"
    },
    "papermill": {
     "duration": 0.004053,
     "end_time": "2025-06-30T02:08:43.229964",
     "exception": false,
     "start_time": "2025-06-30T02:08:43.225911",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "lib = '2'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "97377090",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T02:08:43.233186Z",
     "iopub.status.busy": "2025-06-30T02:08:43.233063Z",
     "iopub.status.idle": "2025-06-30T02:22:22.902123Z",
     "shell.execute_reply": "2025-06-30T02:22:22.901508Z"
    },
    "papermill": {
     "duration": 819.672051,
     "end_time": "2025-06-30T02:22:22.903381",
     "exception": false,
     "start_time": "2025-06-30T02:08:43.231330",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455940 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 455940 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 454379 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 454178 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "587745 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 587745 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 586075 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 583402 UMIs remaining (99.26%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC002\n",
      "115394 UMIs found\n",
      "Collapsing UMIs across probes, 115394 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 115015 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 114456 UMIs remaining (99.19%)\n",
      "145240 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 145240 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 144879 UMIs remaining (99.75%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 121362 UMIs remaining (83.56%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC003\n",
      "38761 UMIs found\n",
      "Collapsing UMIs across probes, 38761 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 38617 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 38465 UMIs remaining (99.24%)\n",
      "53546 UMIs found\n",
      "Collapsing UMIs across probes, 53546 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 53391 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 53049 UMIs remaining (99.07%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC004\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "353446 UMIs found\n",
      "Collapsing UMIs across probes, 353446 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 352206 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 349922 UMIs remaining (99.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "482959 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 482959 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 481617 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 468882 UMIs remaining (97.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC005\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1170446 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1170446 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1166004 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1159525 UMIs remaining (99.07%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1166210 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1166210 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1162784 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1147898 UMIs remaining (98.43%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC006\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1154939 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1154939 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1150414 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1147145 UMIs remaining (99.33%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1131641 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1131641 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1128366 UMIs remaining (99.71%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1115101 UMIs remaining (98.54%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC007\n",
      "104353 UMIs found\n",
      "Collapsing UMIs across probes, 104353 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 103998 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 94232 UMIs remaining (90.30%)\n",
      "132658 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 132658 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 132336 UMIs remaining (99.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 122885 UMIs remaining (92.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC008\n",
      "93523 UMIs found\n",
      "Collapsing UMIs across probes, 93523 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 93165 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 92716 UMIs remaining (99.14%)\n",
      "141999 UMIs found\n",
      "Collapsing UMIs across probes, 141999 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 141627 UMIs remaining (99.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 137261 UMIs remaining (96.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC009\n",
      "59872 UMIs found\n",
      "Collapsing UMIs across probes, 59872 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 59669 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 52090 UMIs remaining (87.00%)\n",
      "109965 UMIs found\n",
      "Collapsing UMIs across probes, 109965 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 109655 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 94662 UMIs remaining (86.08%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC010\n",
      "165877 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 165877 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 165261 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 127083 UMIs remaining (76.61%)\n",
      "209341 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 209341 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 208749 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 177756 UMIs remaining (84.91%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC011\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1685834 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1685834 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1677744 UMIs remaining (99.52%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1412024 UMIs remaining (83.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1603023 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1603023 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1597109 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1477917 UMIs remaining (92.20%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC012\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "402426 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 402426 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 400782 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 356506 UMIs remaining (88.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "480812 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 480812 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 479242 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 441774 UMIs remaining (91.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC013\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2108039 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2108039 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2098022 UMIs remaining (99.52%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1705153 UMIs remaining (80.89%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2082358 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2082358 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2074373 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1919410 UMIs remaining (92.17%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC014\n",
      "156687 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 156687 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 156108 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 139103 UMIs remaining (88.78%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273539 UMIs found\n",
      "Collapsing UMIs across probes, 273539 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 272799 UMIs remaining (99.73%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 254066 UMIs remaining (92.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC015\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2246870 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2246870 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2236718 UMIs remaining (99.55%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2230276 UMIs remaining (99.26%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2207024 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2207024 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2199227 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2181996 UMIs remaining (98.87%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC016\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2102281 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2102281 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2091280 UMIs remaining (99.48%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2090496 UMIs remaining (99.44%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2095903 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2095903 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2087422 UMIs remaining (99.60%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2076505 UMIs remaining (99.07%)\n"
     ]
    }
   ],
   "source": [
    "### first get probe_reads to use for the patient\n",
    "\n",
    "for BC in ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']:\n",
    "    print('Processing BC:', BC)\n",
    "    adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "    gf_dirs = {}\n",
    "    gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    \n",
    "    min_percent_supporting = 0.9\n",
    "    collapse_across_probes = True\n",
    "    for key,gf_dir in gf_dirs.items():\n",
    "        if key == 0:\n",
    "            probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "        else:\n",
    "            probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "    # alt_gapfill_set = pd.read_csv('likelihood_tables/' + BC + '_alt_gapfill_set.csv')\n",
    "    alt_gapfill_set = pd.read_csv('likelihood_tables/all_gapfill_features_filtered_combined.csv')\n",
    "    if len(alt_gapfill_set) > 0:\n",
    "        ## reformat alt gapfill set\n",
    "        manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "        probe_name_to_probe_idx = dict(zip(manifest['name'], manifest['index']))\n",
    "        alt_gapfill_set['probe_idx'] = alt_gapfill_set['name'].map(probe_name_to_probe_idx)\n",
    "        \n",
    "        alt_gapfill_set = pd.concat([\n",
    "            alt_gapfill_set[['probe_idx', 'name', 'gapfill']].rename(columns={'gapfill': 'gapfill_value'}).assign(variant=True),\n",
    "            alt_gapfill_set[['probe_idx', 'name', 'gapfill_from_transcriptome']].rename(columns={'gapfill_from_transcriptome': 'gapfill_value'}).assign(variant=False)\n",
    "        ], ignore_index=True)\n",
    "\n",
    "        alt_gapfill_set = alt_gapfill_set.dropna(subset='probe_idx')\n",
    "        # print(alt_gapfill_set)\n",
    "        \n",
    "        ### from now on, we only need to consider probes that have an alt gapfill\n",
    "        probe_reads = probe_reads.loc[probe_reads['probe_idx'].isin(alt_gapfill_set['probe_idx'])]\n",
    "\n",
    "        unique_gapfill_list = probe_reads[['probe_idx','gapfill']].drop_duplicates()\n",
    "\n",
    "        possible_alleles = {}\n",
    "        for probe_idx in probe_reads['probe_idx'].unique():\n",
    "            possible_alleles[probe_idx] = alt_gapfill_set.loc[alt_gapfill_set['probe_idx'] == probe_idx, 'gapfill_value'].unique()\n",
    "\n",
    "        unique_gapfill_list['likelihood'] = unique_gapfill_list.apply(\n",
    "            lambda row: get_gapfill_likelihoods(row['gapfill'], error_rate_dicts, possible_alleles[row['probe_idx']]), axis=1\n",
    "        )\n",
    "\n",
    "        likelihood_keys = list(set(key for likelihood in unique_gapfill_list['likelihood'] for key in likelihood.keys()))\n",
    "        for key in likelihood_keys:\n",
    "            unique_gapfill_list[key] = unique_gapfill_list['likelihood'].apply(lambda x: x.get(key, None))\n",
    "\n",
    "        unique_gapfill_list = unique_gapfill_list.drop(columns=['likelihood'])\n",
    "\n",
    "        unique_gapfill_list.to_csv('likelihood_tables/' + BC + '_gapfill_likelihoods.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079b59b8",
   "metadata": {
    "papermill": {
     "duration": 0.005145,
     "end_time": "2025-06-30T02:22:22.917247",
     "exception": false,
     "start_time": "2025-06-30T02:22:22.912102",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 824.781138,
   "end_time": "2025-06-30T02:22:23.939155",
   "environment_variables": {},
   "exception": null,
   "input_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "output_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-06-30T02:08:39.158017",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}