{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00ee2077",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:52:39.898903Z",
     "iopub.status.busy": "2025-06-30T01:52:39.898730Z",
     "iopub.status.idle": "2025-06-30T01:52:43.107077Z",
     "shell.execute_reply": "2025-06-30T01:52:43.106671Z"
    },
    "papermill": {
     "duration": 3.212578,
     "end_time": "2025-06-30T01:52:43.108276",
     "exception": false,
     "start_time": "2025-06-30T01:52:39.895698",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "61e904c4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:52:43.115490Z",
     "iopub.status.busy": "2025-06-30T01:52:43.115198Z",
     "iopub.status.idle": "2025-06-30T01:52:43.118425Z",
     "shell.execute_reply": "2025-06-30T01:52:43.118118Z"
    },
    "papermill": {
     "duration": 0.005804,
     "end_time": "2025-06-30T01:52:43.118955",
     "exception": false,
     "start_time": "2025-06-30T01:52:43.113151",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def get_gapfill_likelihoods(gapfill,error_rate_dicts,alleles):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    likelihood_list = {}\n",
    "    for allele in alleles:\n",
    "        if allele in likelihoods:\n",
    "            likelihood_list[allele] = likelihoods[allele]\n",
    "        else:\n",
    "            likelihood_list[allele] = 10e-10\n",
    "    return likelihood_list\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1cb36f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:52:43.122406Z",
     "iopub.status.busy": "2025-06-30T01:52:43.122163Z",
     "iopub.status.idle": "2025-06-30T01:52:43.131701Z",
     "shell.execute_reply": "2025-06-30T01:52:43.131390Z"
    },
    "papermill": {
     "duration": 0.011826,
     "end_time": "2025-06-30T01:52:43.132220",
     "exception": false,
     "start_time": "2025-06-30T01:52:43.120394",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "alt_gapfill_set1 = pd.read_csv('../part1/likelihood_tables/all_gapfill_features_filtered.csv')\n",
    "alt_gapfill_set2 = pd.read_csv('../part2/likelihood_tables/all_gapfill_features_filtered.csv')\n",
    "alt_gapfill_set_combined = pd.concat([alt_gapfill_set1, alt_gapfill_set2])\n",
    "alt_gapfill_set_combined = alt_gapfill_set_combined.sort_values('likelihood_wt_given_observed').drop_duplicates(subset='name')\n",
    "\n",
    "alt_gapfill_set_combined.to_csv('likelihood_tables/all_gapfill_features_filtered_combined.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b3de81c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:52:43.135996Z",
     "iopub.status.busy": "2025-06-30T01:52:43.135833Z",
     "iopub.status.idle": "2025-06-30T01:52:43.137682Z",
     "shell.execute_reply": "2025-06-30T01:52:43.137393Z"
    },
    "papermill": {
     "duration": 0.004011,
     "end_time": "2025-06-30T01:52:43.138194",
     "exception": false,
     "start_time": "2025-06-30T01:52:43.134183",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "lib = '1'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55db8a1f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:52:43.142613Z",
     "iopub.status.busy": "2025-06-30T01:52:43.142496Z",
     "iopub.status.idle": "2025-06-30T02:08:37.578414Z",
     "shell.execute_reply": "2025-06-30T02:08:37.577803Z"
    },
    "papermill": {
     "duration": 954.438807,
     "end_time": "2025-06-30T02:08:37.579547",
     "exception": false,
     "start_time": "2025-06-30T01:52:43.140740",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC001\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "568648 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 568648 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 566830 UMIs remaining (99.68%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 553555 UMIs remaining (97.35%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "674061 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 674061 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 671947 UMIs remaining (99.69%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 655189 UMIs remaining (97.20%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC002\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "387979 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 387979 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 386424 UMIs remaining (99.60%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 366951 UMIs remaining (94.58%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "447368 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 447368 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 445728 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 417432 UMIs remaining (93.31%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC003\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "616887 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 616887 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 614468 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 593738 UMIs remaining (96.25%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "720635 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 720635 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 717840 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 687066 UMIs remaining (95.34%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC004\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "917786 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 917786 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 914057 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 852264 UMIs remaining (92.86%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1111593 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1111593 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1107432 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1020593 UMIs remaining (91.81%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC005\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1419549 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1419549 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1414497 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1364100 UMIs remaining (96.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1645255 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1645255 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1639748 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1580943 UMIs remaining (96.09%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC006\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1994670 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1994670 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1987444 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1861145 UMIs remaining (93.31%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2261733 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2261733 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2254124 UMIs remaining (99.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2089567 UMIs remaining (92.39%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC007\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1756198 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1756198 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1749625 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1613208 UMIs remaining (91.86%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2010873 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2010873 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2003477 UMIs remaining (99.63%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1828215 UMIs remaining (90.92%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC008\n",
      "45584 UMIs found\n",
      "Collapsing UMIs across probes, 45584 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 45445 UMIs remaining (99.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 44990 UMIs remaining (98.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55256 UMIs found\n",
      "Collapsing UMIs across probes, 55256 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 55122 UMIs remaining (99.76%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 40745 UMIs remaining (73.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC009\n",
      "58081 UMIs found\n",
      "Collapsing UMIs across probes, 58081 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 57917 UMIs remaining (99.72%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 57091 UMIs remaining (98.30%)\n",
      "74056 UMIs found\n",
      "Collapsing UMIs across probes, 74056 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 73835 UMIs remaining (99.70%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 65548 UMIs remaining (88.51%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC010\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2292046 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2292046 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2281933 UMIs remaining (99.56%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2106774 UMIs remaining (91.92%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2796720 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2796720 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2785174 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 2571310 UMIs remaining (91.94%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC011\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1138142 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1138142 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1133722 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1054189 UMIs remaining (92.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1271503 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1271503 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1266866 UMIs remaining (99.64%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1082304 UMIs remaining (85.12%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC012\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1345020 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1345020 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1339466 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1288440 UMIs remaining (95.79%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1558831 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1558831 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1552968 UMIs remaining (99.62%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1480120 UMIs remaining (94.95%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC013\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1450740 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1450740 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1445663 UMIs remaining (99.65%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1401531 UMIs remaining (96.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1651562 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1651562 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1646034 UMIs remaining (99.67%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1587987 UMIs remaining (96.15%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC014\n",
      "139217 UMIs found\n",
      "Collapsing UMIs across probes, 139217 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 138856 UMIs remaining (99.74%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 132953 UMIs remaining (95.50%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "179556 UMIs found\n",
      "Collapsing UMIs across probes, 179556 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 179076 UMIs remaining (99.73%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 171766 UMIs remaining (95.66%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC015\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1729224 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1729224 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1722106 UMIs remaining (99.59%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1611729 UMIs remaining (93.21%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2009030 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 2009030 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 2001147 UMIs remaining (99.61%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1857143 UMIs remaining (92.44%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing BC: BC016\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1368673 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1368673 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1361869 UMIs remaining (99.50%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1298549 UMIs remaining (94.88%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1630738 UMIs found\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collapsing UMIs across probes, 1630738 UMIs remaining (100.00%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 1623116 UMIs remaining (99.53%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering cells based on min counts (0) and genes (0) in WTA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering probe reads based on cell barcodes in adata, 1550467 UMIs remaining (95.08%)\n"
     ]
    }
   ],
   "source": [
    "### first get probe_reads to use for the patient\n",
    "\n",
    "for BC in ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']:\n",
    "    print('Processing BC:', BC)\n",
    "    adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "    gf_dirs = {}\n",
    "    gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    \n",
    "    min_percent_supporting = 0.9\n",
    "    collapse_across_probes = True\n",
    "    for key,gf_dir in gf_dirs.items():\n",
    "        if key == 0:\n",
    "            probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "        else:\n",
    "            probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "    # alt_gapfill_set = pd.read_csv('likelihood_tables/' + BC + '_alt_gapfill_set.csv')\n",
    "    alt_gapfill_set = pd.read_csv('likelihood_tables/all_gapfill_features_filtered_combined.csv')\n",
    "    if len(alt_gapfill_set) > 0:\n",
    "        ## reformat alt gapfill set\n",
    "        manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "        probe_name_to_probe_idx = dict(zip(manifest['name'], manifest['index']))\n",
    "        alt_gapfill_set['probe_idx'] = alt_gapfill_set['name'].map(probe_name_to_probe_idx)\n",
    "        \n",
    "        alt_gapfill_set = pd.concat([\n",
    "            alt_gapfill_set[['probe_idx', 'name', 'gapfill']].rename(columns={'gapfill': 'gapfill_value'}).assign(variant=True),\n",
    "            alt_gapfill_set[['probe_idx', 'name', 'gapfill_from_transcriptome']].rename(columns={'gapfill_from_transcriptome': 'gapfill_value'}).assign(variant=False)\n",
    "        ], ignore_index=True)\n",
    "\n",
    "        alt_gapfill_set = alt_gapfill_set.dropna(subset='probe_idx')\n",
    "        # print(alt_gapfill_set)\n",
    "        \n",
    "        ### from now on, we only need to consider probes that have an alt gapfill\n",
    "        probe_reads = probe_reads.loc[probe_reads['probe_idx'].isin(alt_gapfill_set['probe_idx'])]\n",
    "\n",
    "        unique_gapfill_list = probe_reads[['probe_idx','gapfill']].drop_duplicates()\n",
    "\n",
    "        possible_alleles = {}\n",
    "        for probe_idx in probe_reads['probe_idx'].unique():\n",
    "            possible_alleles[probe_idx] = alt_gapfill_set.loc[alt_gapfill_set['probe_idx'] == probe_idx, 'gapfill_value'].unique()\n",
    "\n",
    "        unique_gapfill_list['likelihood'] = unique_gapfill_list.apply(\n",
    "            lambda row: get_gapfill_likelihoods(row['gapfill'], error_rate_dicts, possible_alleles[row['probe_idx']]), axis=1\n",
    "        )\n",
    "\n",
    "        likelihood_keys = list(set(key for likelihood in unique_gapfill_list['likelihood'] for key in likelihood.keys()))\n",
    "        for key in likelihood_keys:\n",
    "            unique_gapfill_list[key] = unique_gapfill_list['likelihood'].apply(lambda x: x.get(key, None))\n",
    "\n",
    "        unique_gapfill_list = unique_gapfill_list.drop(columns=['likelihood'])\n",
    "\n",
    "        unique_gapfill_list.to_csv('likelihood_tables/' + BC + '_gapfill_likelihoods.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f142d9b0",
   "metadata": {
    "papermill": {
     "duration": 0.005691,
     "end_time": "2025-06-30T02:08:37.592773",
     "exception": false,
     "start_time": "2025-06-30T02:08:37.587082",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 959.734547,
   "end_time": "2025-06-30T02:08:38.615150",
   "environment_variables": {},
   "exception": null,
   "input_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "output_path": "step4_save_gapfill_likelihoods_in_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-06-30T01:52:38.880603",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}