{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c27a3cfc",
   "metadata": {
    "papermill": {
     "duration": 0.001947,
     "end_time": "2025-06-30T01:07:48.910192",
     "exception": false,
     "start_time": "2025-06-30T01:07:48.908245",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Below saves the frequencies and likelihoods of all gapfills for all patients. Originally this only saved non-wildtype but it is informative to keep numbers of all.\n",
    "This strategy could be generic, although likelihoods should be updated with empirical frequencies when available (i.e. if those probes have been used on normal tissue or cell lines; in this case they have not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "54b95d4a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:48.914233Z",
     "iopub.status.busy": "2025-06-30T01:07:48.914057Z",
     "iopub.status.idle": "2025-06-30T01:07:51.942164Z",
     "shell.execute_reply": "2025-06-30T01:07:51.941758Z"
    },
    "papermill": {
     "duration": 3.03121,
     "end_time": "2025-06-30T01:07:51.943193",
     "exception": false,
     "start_time": "2025-06-30T01:07:48.911983",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "from gapfill_utilities import utils as gf_utils\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os\n",
    "sys.path.append('/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/src/')\n",
    "import utils as lm_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c693951b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.947410Z",
     "iopub.status.busy": "2025-06-30T01:07:51.947129Z",
     "iopub.status.idle": "2025-06-30T01:07:51.960090Z",
     "shell.execute_reply": "2025-06-30T01:07:51.959790Z"
    },
    "papermill": {
     "duration": 0.015396,
     "end_time": "2025-06-30T01:07:51.960622",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.945226",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘unexpected_gapfill_likelihoods’: File exists\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "256"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_wt_likelihood(gapfill,error_rate_dicts,wt_gapfill):\n",
    "    likelihoods = lm_utils.get_likelihoods_of_true_allele(gapfill, error_rate_dicts)\n",
    "    if wt_gapfill in likelihoods:\n",
    "        return likelihoods[wt_gapfill]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "error_rate_dir = '/home/blattms1/projects/gapfill/4CL_JAK2/likelihood_model/error_rate_dicts_HB'\n",
    "error_rate_dicts = lm_utils.get_error_rate_dicts(error_rate_dir)\n",
    "\n",
    "os.system('mkdir unexpected_gapfill_likelihoods')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6dfbf134",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.964307Z",
     "iopub.status.busy": "2025-06-30T01:07:51.964093Z",
     "iopub.status.idle": "2025-06-30T01:07:51.965925Z",
     "shell.execute_reply": "2025-06-30T01:07:51.965630Z"
    },
    "papermill": {
     "duration": 0.004256,
     "end_time": "2025-06-30T01:07:51.966455",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.962199",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "lib = '1'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8023af43",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-06-30T01:07:51.970094Z",
     "iopub.status.busy": "2025-06-30T01:07:51.969958Z",
     "iopub.status.idle": "2025-06-30T01:27:06.595126Z",
     "shell.execute_reply": "2025-06-30T01:27:06.594503Z"
    },
    "papermill": {
     "duration": 1154.628246,
     "end_time": "2025-06-30T01:27:06.596177",
     "exception": false,
     "start_time": "2025-06-30T01:07:51.967931",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "568648 UMIs found\n",
      "Collapsing UMIs across probes, 568648 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 566830 UMIs remaining (99.68%)\n",
      "Filtering cells based on min counts (0) and genes (0) in WTA\n",
      "Filtering probe reads based on cell barcodes in adata, 552672 UMIs remaining (97.19%)\n",
      "674061 UMIs found\n",
      "Collapsing UMIs across probes, 674061 UMIs remaining (100.00%)\n",
      "Filtering probe reads based on read threshold (0) and min percent supporting (0.9), 671947 UMIs remaining (99.69%)\n",
      "Filtering cells based on min counts (0) and genes (0) in WTA\n",
      "Filtering probe reads based on cell barcodes in adata, 650774 UMIs remaining (96.55%)\n"
     ]
    }
   ],
   "source": [
    "### first get probe_reads to use for the patient\n",
    "\n",
    "# for BC in ['BC001', 'BC002', 'BC003', 'BC004', 'BC005', 'BC006', 'BC007', 'BC008', 'BC009', 'BC010', 'BC011', 'BC012', 'BC013', 'BC014', 'BC015', 'BC016']:\n",
    "for BC in ['BC001']:\n",
    "    adata_path = '/data1/lareauc/users/blattms/gapfill/wta_adata/20250605_MPN_' + lib + '_' + BC + '.h5ad'\n",
    "\n",
    "    gf_dirs = {}\n",
    "    gf_dirs[0] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_1/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "    gf_dirs[1] = '/data1/lareauc/projects/gapfill/analysis/20250605_MPN_16plex/gf_MPN_16plex_part' + lib + '_2/BC' + str(int(BC.replace('BC',''))) + '_giftwrap/'\n",
    "\n",
    "    min_percent_supporting = 0.9\n",
    "    collapse_across_probes = True\n",
    "    for key,gf_dir in gf_dirs.items():\n",
    "        if key == 0:\n",
    "            probe_reads = gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)\n",
    "        else:\n",
    "            probe_reads = pd.concat([probe_reads, gf_utils.get_input_probe_reads(gf_dir, read_threshold = 0, cell_barcode_suffix = '-' + str(key), adata_path = adata_path, min_percent_supporting=min_percent_supporting, collapse_across_probes=collapse_across_probes)], ignore_index=True)\n",
    "\n",
    "    exclude_keyword = '0bp'\n",
    "    manifest = gf_utils.get_manifest(gf_dirs[0])\n",
    "    if 'gapfill_from_transcriptome' not in probe_reads.columns:\n",
    "        probe_reads = probe_reads.merge(manifest[['gapfill_from_transcriptome','name']], left_on='probe_idx',right_index=True,how='left')\n",
    "    probe_reads['gapfill'] = probe_reads['gapfill'].fillna('')\n",
    "    probe_reads['gapfill_from_transcriptome'] = probe_reads['gapfill_from_transcriptome'].fillna('')\n",
    "    if probe_reads['name'].isna().any():\n",
    "        raise ValueError('Probes with no name found in probe reads for ' + BC)\n",
    "    probe_reads = probe_reads.loc[probe_reads['name'].str.contains(exclude_keyword) == False]\n",
    "\n",
    "    counts_in_library = len(probe_reads)\n",
    "    gapfill_frequencies = probe_reads.groupby('probe_idx')['gapfill'].value_counts(normalize=True).rename('frequency').reset_index()\n",
    "    gapfill_counts = probe_reads.groupby('probe_idx')['gapfill'].value_counts().rename('count_of_this_gapfill').reset_index()\n",
    "    probe_counts = probe_reads.groupby('probe_idx').size().rename('count_of_this_probe').reset_index()\n",
    "\n",
    "    to_resolve = probe_reads[['probe_idx','name','gapfill','gapfill_from_transcriptome']].drop_duplicates()\n",
    "    to_resolve['likelihood'] = to_resolve.apply(\n",
    "        lambda row: get_wt_likelihood(row['gapfill'], error_rate_dicts, row['gapfill_from_transcriptome']), axis=1\n",
    "    )\n",
    "    to_resolve = to_resolve.merge(gapfill_frequencies, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(gapfill_counts, on = ['probe_idx','gapfill'], how='left')\n",
    "    to_resolve = to_resolve.merge(probe_counts, on = ['probe_idx'], how='left')\n",
    "\n",
    "    to_add = manifest.loc[~(manifest.index.isin(to_resolve['probe_idx'])) & (manifest['name'].str.contains(exclude_keyword) == False)].reset_index().rename(columns={'index': 'probe_idx'})\n",
    "    to_resolve = pd.concat([to_resolve, to_add[['probe_idx','name','gapfill_from_transcriptome']]]).reset_index(drop=True)\n",
    "    to_resolve[['count_of_this_gapfill','count_of_this_probe']] = to_resolve[['count_of_this_gapfill','count_of_this_probe']].fillna(0)\n",
    "    to_resolve.to_csv('unexpected_gapfill_likelihoods/unresolved_gapfills_' + BC + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6354e4e",
   "metadata": {
    "papermill": {
     "duration": 0.00586,
     "end_time": "2025-06-30T01:27:06.609289",
     "exception": false,
     "start_time": "2025-06-30T01:27:06.603429",
     "status": "completed"
    },
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4161a58",
   "metadata": {
    "papermill": {
     "duration": 0.005621,
     "end_time": "2025-06-30T01:27:06.620298",
     "exception": false,
     "start_time": "2025-06-30T01:27:06.614677",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gapfill_utilities",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1159.706566,
   "end_time": "2025-06-30T01:27:07.641859",
   "environment_variables": {},
   "exception": null,
   "input_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "output_path": "step2_save_gapfill_frequencies_for_feature_set.ipynb",
   "parameters": {},
   "start_time": "2025-06-30T01:07:47.935293",
   "version": "2.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
